### ISSUE TITLE
Dynlink version/interface mismatch errors should report OCaml versions

### ISSUE THREAD

**Issue by JasonGross** [ocaml/ocaml#9532]
In using Coq plugins, the most frequent source of errors like
```
Error: while loading
/home/fiat/plv/fiat-crypto/rewriter/src/Rewriter/Util/plugins/strategy_tactic_plugin.cmxs:
implementation mismatch on Stdlib__list
```
is when the version of OCaml used to compile the cmxs file does not match the version of OCaml used to compile Coq.  I think it'd be a great help if, when the interfaces mismatch either due to differing OCaml versions or differing OCaml compiler flags, the error message reported this as a hint of the form
```
This may be due to the fact that $file was compiled with OCaml version $version [or flags $flags], which is incompatible with OCaml version $current_version [or flags $current_flags]
```

See also https://github.com/coq/coq/issues/12232, https://github.com/coq/coq/pull/12244#issuecomment-623193948

In particular, this is useful for users of Coq developments who are not well-versed in the OCaml ecosystem, and are thus using Coq OCaml plugins without having ever written any OCaml themselves.

**Comment by dra27** [comment]
Improvements have been made in this direction in #463.

**Comment by dra27** [comment]
How common is it amongst Coq users to see a plugin compiled in a *newer* version of OCaml loaded by an older one (i.e. the hardest case for improving the error), or is mismatched flags and/or forgetting to recompile an old plugin much more common when this happens?

**Comment by ejgallego** [comment]
@dra27 I think most of the time Coq users see this because their build setup is broken; there is a long tradition in Coq land to have a large build mess; so I dunno how useful this would be.

In my opinion we should rather encourage people to use saner build setups.

**Comment by JasonGross** [comment]
@dra27 I believe mismatched flags is relatively uncommon.  I believe the most common case is that the version used to compile the plugin is newer (for example, this happened on Arch recently because the Arch devs upgraded their OCaml package without updating/recompiling their Coq package, so plugins built by users used a newer OCaml with an older Coq.  Even if this is a case of "broken build system", having a good error message here would have saved time in debugging this issue and reporting it upstream).  I believe the other situation is a bit more rare, but still possible.  (I think it's also fine if sometimes exact version numbers can't be reported, and all that can be said is "a version of OCaml incompatible with version foo".)

@ejgallego I don't think that's a helpful perspective, even if the most common cause of this message is indeed broken build systems.  Not all versions of Coq are available in opam (for example, sometimes I'm asked to test out a PR on a development) and anyway it is untrainable to assume that all users will use opam to manage all their versions of Coq.  Especially because `opam switch` is a global operation, and sometimes I need to run two versions of Coq in parallel (for example, to minimize a bug), and not all versions of Coq support the same versions of OCaml (and sometimes I'm asked to test a particular version of Coq with a particular version of OCaml, like with the recent segfaults on Windows).  So if I switch which version of OCaml I'm using globally and forget to recompile Coq, or I neglect to switch back to an older version when I want to run some code that only works in Coq 8.4, I might see this error message.  Having it report that the versions mismatch would save me time in these circumstances, when the cause of it is not in my mental cache.

**Comment by JasonGross** [comment]
> Improvements have been made in this direction in #463.

Ah, neat!  Does that apply also to dynlink?  And what version of OCaml is it for?  If the error message now mentions version mismatch on both older and newer versions of OCaml, and names the version it's expecting, that's sufficient for me.

**Comment by ejgallego** [comment]
@JasonGross my point is that version information does little for your problem, which is one of mismatched interfaces.

**Comment by ejgallego** [comment]
> I need to run two versions of Coq in parallel (for example, to minimize a bug), and not all versions of Coq support the same versions of OCaml (and sometimes I'm asked to test a particular version of Coq with a particular version of OCaml, like with the recent segfaults on Windows). So if I switch which version of OCaml I'm using globally and forget to recompile Coq, or I neglect to switch back to an older version when I want to run some code that only works in Coq 8.4, I might see this error message.

There is a clear issue with your procedure here, and indeed it is that your workflow relies on manual steps, then it is not a surprise package consistency is broken.

**Comment by gasche** [comment]
The work on magic numbers was at first only for ocamlobjinfo (not the compiler, a separate helper tool), and is slowly getting used in various parts of the compiler, but I don't think that .cmxs are covered yet. Note that it only applies to incompatibility in magic numbers; cmxs have a magic number as well, and I think that it should now be bumped at each release (in the past it may have been bumped less often when the representations were compatible), but the error message @JasonGross reports

```
Error: while loading
/home/fiat/plv/fiat-crypto/rewriter/src/Rewriter/Util/plugins/strategy_tactic_plugin.cmxs:
implementation mismatch on Stdlib__list
```

is *not* a magic-number mismatch, so there would be no better error message printed here.

I am in fact not sure how you got this specific error message. It seems that the magic numbers are compatible, that the signatures for the standard library List modules are compatible, but that the implementations are not. (So maybe a version change without cmxs-number bump and List interface change, or maybe a code-production difference like flambda/non-flambda.) Is it really representative of the errors you see in the wild?

**Comment by JasonGross** [comment]
@ejgallego Knowing that the issue is OCaml versions is significantly more useful than just knowing that the interfaces are mismatched (a priori I have no idea why they might be mismatched because I never interact with the interface of Stdlib__list, while I do interact with OCaml versions).  Knowing what the expected OCaml version is can help me track down where the version mismatch is coming from, though it's probably not all that important.  Similarly, knowing the version found can also help me track down the issue, but, again, it's not as big an improvement as having it report that the issue is a version one.

**Comment by ejgallego** [comment]
@JasonGross as pointed out by Gabriel in the case you mention actually reporting the OCaml version message would likely be incorrect.

**Comment by JasonGross** [comment]
> I am in fact not sure how you got this specific error message. It seems that the magic numbers are compatible, that the signatures for the standard library List modules are compatible, but that the implementations are not. (So maybe a version change without cmxs-number bump and List interface change, or maybe a code-production difference like flambda/non-flambda.) Is it really representative of the errors you see in the wild?

This is the error message @andres-erbsen got recently on Arch, in the "Arch updated their OCaml package without updating their Coq package" situation.  @andres-erbsen, do you have access to the OCaml versions/flags of the packages before and after the update?
(When he asked me, my first thought was outdated cmxs files, and it took me a couple of minutes before remembering this could be caused by a mismatch between the version of OCaml that compiled Coq and the version present on the system.)

**Comment by andres-erbsen** [comment]
I don't have a snapshot from before the upgrade because I only found out about this when I tried to use a non-bundled Coq plugin for the first time in a while. The current Arch packages are still unusable for building plugins, I believe. [ocaml](https://archive.is/BeB0T) built on 2020-03-19, [coq](https://archive.is/psBOU) built on 2020-01-28. The out-of-date flag on coq is by me, for 8.11.1.

**Comment by JasonGross** [comment]
Based on [the changelog](https://git.archlinux.org/svntogit/packages.git/log/trunk?h=packages/ocaml), which says:

> Age | Commit message (Expand) | Author
> -- | -- | --
> 2020-03-19 | upgpkg: ocaml 4.09.1-1 | felixonmars
> 2019-11-27 | use a more reliable method to install the compiler-libs, Fixes #64379 | juergen
> 2019-10-10 | upgpkg: ocaml 4.09.0-1 | juergen

it looks like this is the error message from trying to use a Coq built with 4.09.0 when the system OCaml is 4.09.1.
